File: Fahrschule/Classes/LicenseClassSelectViewController.m
//
//  LicenseClassSelectViewController.m
//  Fahrschule
//
//  Created by Johan Olsson on 20.04.11.
//  Copyright 2011 freenet. All rights reserved.
//

#import "LicenseClassSelectViewController.h"
//#import "GroupedCellBackgroundView.h"
//#import "LicenseClassTableViewCell.h"
//#import "UIColor+ColorWithHexString.h"


@implementation LicenseClassSelectViewController

@synthesize licenseClassesDict=_licenseClassesDict;
@synthesize keyList=_keyList;
@synthesize tableView=_tableView;

NSInteger intSort(id num1, id num2, void *context)
{
    int v1 = [num1 intValue];
    int v2 = [num2 intValue];
    if (v1 < v2)
        return NSOrderedAscending;
    else if (v1 > v2)
        return NSOrderedDescending;
    else
        return NSOrderedSame;
}

//- (id)initInManagedObjectContext:(NSManagedObjectContext *)context
//{
//    self = [super initInManagedObjectContext:context];
//    if (self) {
//
//        _licenseClassesDict = [[NSDictionary alloc] initWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"LicenseClasses" ofType:@"plist"]];
//        
//        _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStyleGrouped];
//        [self.tableView setBackgroundColor:[UIColor clearColor]];
//        [self.tableView setBackgroundView:nil];
//        [self.tableView setDataSource:self];
//        [self.tableView setDelegate:self];
//        
//        self.keyList = [NSArray arrayWithObjects:@"1", @"2", @"8192", @"16384", @"4", @"8", @"16", @"32", @"64", @"128", @"512", @"1024", @"4096", nil];
//        
//        [self setTitle:NSLocalizedString(@"Führerscheinklasse", @"")];
//        
//        if (self.userSettings.licenseClass != kUnknownLicenseClass && self.userSettings.licenseClass != kLicenseClassS &&
//            self.userSettings.licenseClass != kLicenseClassM && !self.iPad) {
//            UIBarButtonItem *closeButton = [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"Zurück", @"") style:UIBarButtonItemStylePlain target:self action:@selector(closeModalView:)];
//            [self.navigationItem setLeftBarButtonItem:closeButton animated:NO];
//        }
//    }
//    return self;
//}

#pragma mark - View lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.view.autoresizesSubviews = YES;
    
    self.edgesForExtendedLayout = UIRectEdgeNone;
    
    self.navigationController.navigationBar.barStyle = UIBarStyleBlackOpaque;
    
    self.tableView.frame = CGRectMake(0.0, 0.0, self.view.frame.size.width, self.view.frame.size.height);
    self.tableView.autoresizingMask = (UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth);
    
//    if (self.iPad) {
//        self.view.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"bg_mitGitter"]];
//    }
//    else {
//        UIImageView *iv = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"bg_mitGitter"]];
//        iv.frame = CGRectMake(0.0, 0.0, self.view.frame.size.width, self.view.frame.size.height);
//        iv.autoresizingMask = (UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth);
//        [self.view addSubview:iv];
//
//    }
    
    [self.view addSubview:self.tableView];


}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [self.keyList count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    
//    LicenseClassTableViewCell *cell = (LicenseClassTableViewCell *)[tableView dequeueReusableCellWithIdentifier:CellIdentifier];
//    if (cell == nil) {
//        cell = [[LicenseClassTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];
//        cell.textLabel.backgroundColor = [UIColor clearColor];
//        cell.detailTextLabel.backgroundColor = [UIColor clearColor];
//    }
//    
//    GroupedCellBackgroundView *bgView = [[GroupedCellBackgroundView alloc] initWithFrame:CGRectZero];
//    bgView.fillColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"bg_fragenkatalog.png"]];
//    bgView.borderColor = [UIColor lightGrayColor];
//    
//    GroupedCellBackgroundView *selBgView = [[GroupedCellBackgroundView alloc] initWithFrame:CGRectZero];
//    selBgView.fillColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"bg_fragenkatalog_aktiv.png"]];
//    selBgView.borderColor = [UIColor lightGrayColor];
//    
//    if (indexPath.row == 0) {
//        bgView.position = CustomCellBackgroundViewPositionTop;
//        selBgView.position = CustomCellBackgroundViewPositionTop;
//    }
//    else if (indexPath.row == [self.keyList count] - 1) {
//        bgView.position = CustomCellBackgroundViewPositionBottom;
//        selBgView.position = CustomCellBackgroundViewPositionBottom;
//    }
//    else {
//        bgView.position = CustomCellBackgroundViewPositionMiddle;
//        selBgView.position = CustomCellBackgroundViewPositionMiddle;
//    }
//    
//    cell.backgroundView = bgView;
//    cell.selectedBackgroundView = selBgView;
//    
//    NSDictionary *currentLicenseClass = [self.licenseClassesDict objectForKey:[self.keyList objectAtIndex:indexPath.row]];
//    cell.imageView.image = [UIImage imageNamed:[currentLicenseClass objectForKey:@"image"]];
//    cell.textLabel.text = [currentLicenseClass objectForKey:@"title"];
//    cell.detailTextLabel.text = [currentLicenseClass objectForKey:@"desc"];
//    cell.classLabel.text = [currentLicenseClass objectForKey:@"class"];
//    
//    cell.detailTextLabel.font = [UIFont systemFontOfSize:12.0];
//        
//    if ([[self.keyList objectAtIndex:indexPath.row] isEqual:[NSString stringWithFormat:@"%d", self.userSettings.licenseClass]]) {
//        cell.accessoryType = UITableViewCellAccessoryCheckmark;
//    }
//    else {
//        cell.accessoryType = UITableViewCellAccessoryNone;
//    }
//    
//#ifdef FAHRSCHULE_LITE
//    if (indexPath.section == 0 && indexPath.row == 4) {
//        cell.textLabel.textColor = [UIColor blackColor];
//        cell.classLabel.textColor = [UIColor colorWithRGBHex:0x394E75];
//        cell.userInteractionEnabled = YES;
//    }
//    else {
//        cell.textLabel.textColor = [UIColor grayColor];
//        cell.classLabel.textColor = [UIColor grayColor];
//        cell.userInteractionEnabled = NO;
//    }
//#endif
    
    return nil;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
//    if (indexPath.section == 0) {
//        [self.userSettings setLicenseClass:[[self.keyList objectAtIndex:indexPath.row] intValue]];
//        
//        if (self.userSettings.licenseClass == kLicenseClassC1 || self.userSettings.licenseClass == kLicenseClassC || self.userSettings.licenseClass == kLicenseClassCE
//            || self.userSettings.licenseClass == kLicenseClassD1 || self.userSettings.licenseClass == kLicenseClassD) {
//            
//            [self.userSettings setTeachingType:kAdditionalLicense]; 
//        }
//        else if (self.userSettings.licenseClass == kLicenseClassMOFA) {
//            [self.userSettings setTeachingType:kFirstTimeLicense];
//        }
//        
//        [self.tableView reloadData];
//    }
//
//    [self dismissViewControllerAnimated:YES completion:NULL];
}

#pragma mark - Actions

- (void)closeModalView:(id)sender
{
    [self dismissViewControllerAnimated:YES completion:NULL];
}

@end

File: Fahrschule/Classes/LicenseClassSelectViewController.swift
//
//  LicenseClassSelectViewController.swift
//  Fahrschule
//
//  Created on 22.06.15.
//  Copyright (c) 2015. All rights reserved.
//

import UIKit

class LicenseClassSelectViewController: UICollectionViewController, UICollectionViewDelegateFlowLayout {

    var licenseClassesDict:  NSDictionary?
    var keyList = ["1", "2", "8192", "16384", "4", "8", "16", "32", "64", "128", "512", "1024", "4096"]
    let checkedImage = UIImage(named: "bestanden")
    
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.clearsSelectionOnViewWillAppear = false
        let path = NSBundle.mainBundle().pathForResource("LicenseClasses", ofType: "plist")
        self.licenseClassesDict = NSDictionary(contentsOfFile: path!)
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

//    MARK: - Table view data source

    override func  collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.keyList.count
    }
    
    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier("Cell", forIndexPath: indexPath) as! LicenseCell
        
        let licenseClass = self.keyList[indexPath.row]
        let dict: NSDictionary = self.licenseClassesDict?.objectForKey(licenseClass) as! NSDictionary
        cell.titleLabel?.text = dict["title"] as? String
        cell.subtitleLabel?.text = dict["desc"] as? String
        cell.classLabel?.text = dict["class"] as? String
        
        
        let imageName: String = dict["image"] as! String
        let image = UIImage(named: imageName)
        
        if image != nil {
            cell.imageView?.image = image
        } else {
            cell.imageView?.image = nil
            println("\(imageName)")
        }
        
//        let settings = Settings.sharedSettings() as! Settings
//        let lc = settings.licenseClass as LicenseClass
        
        
        if licenseClass.toInt() == Settings.sharedSettings()!.licenseClass.rawValue {
            cell.iconImageView.image = checkedImage
        } else {
            cell.iconImageView.image = nil
        }
        
        
        
        return cell
    }
    
//    MARK: Collection View Delegate
    override func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        
        let settings = Settings.sharedSettings() as! Settings
        let oldValue = settings.licenseClass as LicenseClass
        let newValue: Int = self.keyList[indexPath.item].toInt()!
        if newValue != oldValue.rawValue {
            settings.licenseClass = LicenseClass(rawValue: newValue)!
            
            let cell = collectionView.cellForItemAtIndexPath(indexPath) as! LicenseCell
            cell.iconImageView.image = checkedImage
            
            if let idx = find(self.keyList, String(oldValue.rawValue)) {
                let oldIndexPath = NSIndexPath(forItem: idx, inSection: 0)
                if let oldCell = collectionView.cellForItemAtIndexPath(oldIndexPath) as? LicenseCell  {
                    oldCell.iconImageView.image = nil
                }
            }
                        
            if settings.licenseClass == .LicenseClassC1 || settings.licenseClass == .LicenseClassC || settings.licenseClass == LicenseClass.LicenseClassCE || settings.licenseClass == .LicenseClassD1 || settings.licenseClass == .LicenseClassD {
                
                settings.teachingType = .AdditionalLicense
            }
            else if settings.licenseClass == .LicenseClassMOFA {
                settings.teachingType = .FirstTimeLicense
            }
        }
        
        self.dismissViewControllerAnimated(true, completion: nil)
        
    }
    
//    MARK: - Collection View Flowlayout delegate
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
        let width = UIDevice.currentDevice().userInterfaceIdiom == .Phone ? CGRectGetWidth(collectionView.bounds) : CGRectGetWidth(collectionView.bounds) / 2.0
        return CGSizeMake(width, 44.0)
    }


}
